### PR TITLE
tools: add tool to create compile_commands.json

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
Add a new script, tools/make-compile-commands, which can generate the
compile_commands.json script used by many static analysis tools.

### CODE REVIEW

**Comment by jelly** [comment]
As noted on IRC, this only lacks the `command` key as noted in the [docs](https://clang.llvm.org/docs/JSONCompilationDatabase.html#format). Although it also mentions arguments is generallyy preferred.

**Comment by martinpitt** [comment]
Moving to draft. Jelle had some comments, and ATM this is neither being used, documented, nor smoke-tested.

**Comment by allisonkarlitskaya** [comment]
so, I'm already using this locally to help [ALE](https://github.com/dense-analysis/ale) produce correct warnings for editing C files.

**Comment by allisonkarlitskaya** [comment]
hihi rhel8
```
  GEN      compile_commands.json
  File "./tools/make-compile-commands", line 39
    if match := cc_pattern.fullmatch(line):
              ^
SyntaxError: invalid syntax
```

**Comment by martinpitt** [comment]
```
    with Popen(*popenargs, **kwargs) as process:
TypeError: __init__() got an unexpected keyword argument 'text'
```

File: tools/make-compile-commands [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    with open('Makefile', 'rb') as original_makefile:
+        makefile_copy.write(original_makefile.read())
+
+    # get the build commands
+    output = subprocess.check_output(['make',
+                                      '-f', makefile_copy.name,
+                                      'MAKE=false',   # block recursive make
+                                      'V=1',          # simplify the output
+                                      '--always-make', '--dry-run', 'all-am'],
+                                     text=True)
```
Comment (martinpitt) [inline tools/make-compile-commands:31]: Should be `universal_newlines=True` for backwards compatibility

**Comment by allisonkarlitskaya** [comment]
Interesting thing to note: this introduces a build dependency on Python.  I don't think we have one of those yet, for the tarball: the closest we come is `termschutz`, and this is only used during `make dist`.

**Review by martinpitt — APPROVED** [review]
Thanks! Looking forward to testing it. (I didn't yet, as this is "mostly harmless", and if it works for you, that's great)

**Pull request merged**